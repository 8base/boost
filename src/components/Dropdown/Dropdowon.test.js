// @flow

import * as React from 'react';
import { Dropdown } from './';


describe('<Dropdown />', () => {

  it('should render dropdown', () => {
    const wrapper = mount(
      <Dropdown isOpen={ false }>
        <Dropdown.Head>Head</Dropdown.Head>
        <Dropdown.Body>
          <div className="body">Body</div>
        </Dropdown.Body>
      </Dropdown>,
    );

    expect(wrapper).toMatchSnapshot();
  });


  it('should open and close dropdwon by changing the props', () => {
    const wrapper = mount(
      <Dropdown isOpen={ false }>
        <Dropdown.Head>Head</Dropdown.Head>
        <Dropdown.Body>
          <div className="body">Body</div>
        </Dropdown.Body>
      </Dropdown>,
    );

    expect(wrapper.find(Dropdown.Head).text()).toBe('Head');
    expect(wrapper.find('.body')).toHaveLength(0);

    wrapper.setProps({ isOpen: true });
    expect(wrapper.find('.body')).toHaveLength(1);
    expect(wrapper.find(Dropdown.Body).text()).toBe('Body');

    wrapper.setProps({ isOpen: false });
    expect(wrapper.find('.body')).toHaveLength(0);
  });


  it('should call onClose and onOpen callbacks', () => {
    const onOpenDropdown = jest.fn();
    const onCloseDropdown = jest.fn();

    const wrapper = mount(
      <Dropdown isOpen={ false } onCloseDropdown={ onCloseDropdown } onOpenDropdown={ onOpenDropdown }>
        <Dropdown.Head>Head</Dropdown.Head>
        <Dropdown.Body>
          <div className="body">Body</div>
        </Dropdown.Body>
      </Dropdown>,
    );

    wrapper.find(Dropdown.Head).simulate('click');
    expect(onOpenDropdown).toHaveBeenCalled();

    wrapper.setProps({ isOpen: true });
    wrapper.find(Dropdown.Head).simulate('click');
    expect(onCloseDropdown).toHaveBeenCalled();
  });


  it('should open and close dropdwon by clicking to head', () => {
    const wrapper = mount(
      <Dropdown defaultOpen>
        <Dropdown.Head>Head</Dropdown.Head>
        <Dropdown.Body>
          <div className="body">Body</div>
        </Dropdown.Body>
      </Dropdown>,
    );

    expect(wrapper.find('.body')).toHaveLength(1);
    expect(wrapper.find('.body').text()).toBe('Body');

    wrapper.find(Dropdown.Head).simulate('click');
    expect(wrapper.find('.body')).toHaveLength(0);

    wrapper.find(Dropdown.Head).simulate('click');
    expect(wrapper.find('.body')).toHaveLength(1);
    expect(wrapper.find('.body').text()).toBe('Body');
  });


  it('should render with function in body', () => {
    const onCloseDropdown = jest.fn();

    const wrapper = mount(
      <Dropdown isOpen onCloseDropdown={ onCloseDropdown }>
        <Dropdown.Head>Head</Dropdown.Head>
        <Dropdown.Body stretch>
          { ({ closeDropdown }: *) => (
            <div className="close-dropdown" onClick={ closeDropdown } />
          ) }
        </Dropdown.Body>
      </Dropdown>,
    );


    wrapper.find('.close-dropdown').simulate('click');
    expect(onCloseDropdown).toHaveBeenCalled();
  });
});

