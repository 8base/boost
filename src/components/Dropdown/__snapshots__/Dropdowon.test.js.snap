// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> should render dropdown 1`] = `
<Dropdown
  isOpen={false}
>
  <branch(DropdownPlate)
    isOpen={false}
  >
    <DropdownPlate
      isOpen={false}
    >
      <Boost(dropdown)
        insideRef={[Function]}
        tag="div"
      >
        <Tag
          className="emotion-0 emotion-1"
          insideRef={[Function]}
          tag="div"
          tagName="div"
        >
          <div
            className="emotion-0 emotion-1"
          >
            <Manager
              tag={false}
            >
              <Dropdown.Head>
                <DropdownHead
                  dropdown={
                    Object {
                      "closeDropdown": undefined,
                      "isOpen": false,
                      "openDropdown": undefined,
                      "outsideClickIgnoreClass": "ignore-react-onclickoutside-0",
                      "targetWidth": undefined,
                      "toggleDropdown": [Function],
                    }
                  }
                >
                  <Boost(dropdownHead)
                    className="ignore-react-onclickoutside-0"
                    onClick={[Function]}
                    tagName="div"
                  >
                    <Tag
                      className="ignore-react-onclickoutside-0 emotion-0 emotion-1"
                      onClick={[Function]}
                      tagName="div"
                    >
                      <div
                        className="ignore-react-onclickoutside-0 emotion-0 emotion-1"
                        onClick={[Function]}
                      >
                        <Reference>
                          <InnerReference
                            getReferenceRef={[Function]}
                          >
                            <Boost(dropdownHeadTarget)
                              insideRef={[Function]}
                              tagName="div"
                            >
                              <Tag
                                className="emotion-0 emotion-1"
                                insideRef={[Function]}
                                tagName="div"
                              >
                                <div
                                  className="emotion-0 emotion-1"
                                >
                                  Head
                                </div>
                              </Tag>
                            </Boost(dropdownHeadTarget)>
                          </InnerReference>
                        </Reference>
                      </div>
                    </Tag>
                  </Boost(dropdownHead)>
                </DropdownHead>
              </Dropdown.Head>
              <Dropdown.Body>
                <withProps(OnClickOutside(DropdownBodyBase))
                  dropdown={
                    Object {
                      "closeDropdown": undefined,
                      "isOpen": false,
                      "openDropdown": undefined,
                      "outsideClickIgnoreClass": "ignore-react-onclickoutside-0",
                      "targetWidth": undefined,
                      "toggleDropdown": [Function],
                    }
                  }
                >
                  <OnClickOutside(DropdownBodyBase)
                    dropdown={
                      Object {
                        "closeDropdown": undefined,
                        "isOpen": false,
                        "openDropdown": undefined,
                        "outsideClickIgnoreClass": "ignore-react-onclickoutside-0",
                        "targetWidth": undefined,
                        "toggleDropdown": [Function],
                      }
                    }
                    eventTypes={
                      Array [
                        "mousedown",
                        "touchstart",
                      ]
                    }
                    excludeScrollbar={false}
                    outsideClickIgnoreClass="ignore-react-onclickoutside-0"
                    preventDefault={false}
                    stopPropagation={false}
                  >
                    <DropdownBodyBase
                      background="none"
                      borderRadius="md"
                      closeOnClickOutside={true}
                      disableOnClickOutside={[Function]}
                      dropdown={
                        Object {
                          "closeDropdown": undefined,
                          "isOpen": false,
                          "openDropdown": undefined,
                          "outsideClickIgnoreClass": "ignore-react-onclickoutside-0",
                          "targetWidth": undefined,
                          "toggleDropdown": [Function],
                        }
                      }
                      enableOnClickOutside={[Function]}
                      eventTypes={
                        Array [
                          "mousedown",
                          "touchstart",
                        ]
                      }
                      forceRender={false}
                      isOpen={false}
                      offset="xs"
                      outsideClickIgnoreClass="ignore-react-onclickoutside-0"
                      padding="none"
                      pin="left"
                      placement="bottom"
                      preventDefault={false}
                      stopPropagation={false}
                      withPortal={false}
                    />
                  </OnClickOutside(DropdownBodyBase)>
                </withProps(OnClickOutside(DropdownBodyBase))>
              </Dropdown.Body>
            </Manager>
          </div>
        </Tag>
      </Boost(dropdown)>
    </DropdownPlate>
  </branch(DropdownPlate)>
</Dropdown>
`;
